### PR TITLE
✨ Log in

### PR DESCRIPTION
**Pull request by eliasmalik**
Related to #7

### Changes
* Copy over a lot of stuff from the visitor app
* Use `formik` module to manage form state (instead of doing it ourselves)
* Add the Login (tested), ForgotPassword and ResetPassword pages
* Add a blank Dashboard

### Partnered PR
TwinePlatform/twine-api#370

### Manual Testing https://github.com/TwinePlatform/volunteer-dashboard/issues/10
* Login (as CB_ADMIN: Success)
* Login (as VOLUNTEER_ADMIN: Success)
* Login (Validation error)
* Login (Unrecognised user)
* Forgot/Reset password (Success)
* Forgot password (Unrecognised user)
* Forgot/Reset password (Non-matching passwords)
* Forgot/Reset password (Invalid token)
* Forgot/Reset password (Invalid email)

### Notes
Several thoughts sparked by this PR:
* There's going to be a lot of commonality between the visitor and dashboard apps. We should think of ways to share code between them.
* Having TS now means we can type request and response payloads in the API interface module (we should definitely do this). However, this means we are essentially duplicating types (since they will also be defined in the `twine-api` repo).
* I haven't implemented redux/state-mgmt yet because it's not really necessary at this stage.

### CODE REVIEW

**Comment by eliasmalik** [comment]
I don't really understand why these tests are failing. They pass individually. It seems like the DOM isn't being cleaned up properly, but I haven't been able to isolate it yet.

**Comment by eliascodes** [comment]
I don't really understand why these tests are failing. They pass individually. It seems like the DOM isn't being cleaned up properly, but I haven't been able to isolate it yet.

**Comment by astroash** [comment]
@eliascodes do we know if we want to keep the same style as the visitor app? If so it's worth us setting up a component repo/library.

Altho, all this shared code is making me contemplate whether we should have a mono repo...

Other changes:
- add `.env` instructions to the setup guide
- change default port so that it is different to the visitor app

**Comment by eliasmalik** [comment]
> do we know if we want to keep the same style as the visitor app?

No, we don't, but it's all I can do at the moment.

> If so it's worth us setting up a component repo/library.

Yeah I've wanted to do this. Until now there was no real need. But I think it's better to wait until we have a design, then we will at least be able to see what's common, and which direction (visually) the app is going to be going in.

> Altho, all this shared code is making me contemplate whether we should have a mono repo...

I am increasingly in favour of this, in particular in order to benefit from typing the API. Will have to think about potential down sides.

**Comment by eliasmalik** [comment]
@astroash So I've disabled the tests here until we can figure out wtf is going on with them. Everything is otherwise functional though.

File: src/auth/components/PrivateRoute.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import { Route, RouteProps, Redirect } from 'react-router-dom';
+import { FullScreenBeatLoader } from '../../components/Loaders';
+import { colors } from '../../styles/style_guide';
+import { Roles } from '../../api';
+
+
+/**
+ * Types
+ */
+enum AuthStatus {
```
Comment (astroash) [inline src/auth/components/PrivateRoute.tsx]: Do we want to continue the convention of suffixing enums with `[Name]Enum`

File: src/auth/components/PrivateRoute.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+interface PrivateRouteProps extends RouteProps {}
+
+interface PrivateRouteState {
+  authStatus: AuthStatus;
+}
+
+
+export default class PrivateRoute extends Component<PrivateRouteProps, PrivateRouteState> {
```
Comment (astroash) [inline src/auth/components/PrivateRoute.tsx:25]: We should probably discuss whether we want this project to be written with hooks or class components. I have preference for the former.

File: src/auth/pages/Login/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import styled from 'styled-components';
+import { Link as L, withRouter, RouteComponentProps } from 'react-router-dom';
+import { Grid, Row, Col } from 'react-flexbox-grid';
+import { Form as _Form, FormikActions } from 'formik';
+import { CbAdmins, Response } from '../../../api';
+import LoginForm, { FormValues } from './LoginForm';
+import { H1, H4 } from '../../../components/Headings';
+
+
+/**
```
Comment (astroash) [inline src/auth/pages/Login/Login.tsx:12]: Doesn't really matter, but how come all your comment blocks have an extra `*`

File: src/auth/pages/Login/Login.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+// Submit handler creator
+const createSubmitHandler = (props: LoginProps) =>
+  (values: FormValues, actions: FormikActions<FormValues>) =>
+    CbAdmins.login(values)
+      .then(() => props.history.push('/'))
+      .catch((error) => {
+        const res = error.response;
+
+        if (Response.statusEquals(res, 400)) {
```
Comment (astroash) [inline src/auth/pages/Login/Login.tsx:38]: Do we want an error handler like the visitor app?

File: src/auth/pages/Login/LoginForm.tsx [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+const schema = Joi.object({
+  email: Joi.string().email().required(),
+  password: Joi.string().min(8),
+});
+
+type LoginFormProps = {
+  onSubmit: (v: FormValues, a: FormikActions<FormValues>) => void
+};
+
+
```
Comment (astroash) [inline src/auth/pages/Login/LoginForm.tsx]: Missing Component header comment block

File: src/auth/pages/ResetPassword.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+const schema = Joi.object({
+  password: Joi.string().min(8).required(),
+  passwordConfirm: Joi.string().required().only(Joi.ref('password'))
+    .options({ language: { any: {
+      required: 'password confirmation is required',
+      allowOnly: 'must match password',
+    } } }),
+});
+
```
Comment (astroash) [inline src/auth/pages/ResetPassword.tsx:41]: Missing Component header comment block

File: src/dashboard/Dashboard.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import React from 'react';
+import { RouteComponentProps, withRouter } from 'react-router-dom';
+import { Grid, Row, Col } from 'react-flexbox-grid';
+import { CbAdmins, Response, CommunityBusinesses } from '../api';
+import { redirectOnError } from '../util/routing';
+import NavHeader from '../components/NavHeader/NavHeader';
+
+
```
Comment (astroash) [inline src/dashboard/Dashboard.tsx:8]: Missing header comment blocks

**Review by astroash — CHANGES_REQUESTED** [review]

File: src/auth/pages/Login/Login.tsx (hunk 3 above)
Comment (eliasmalik) [reply src/auth/pages/Login/Login.tsx:12]: That's deliberate, it's a JSDoc convention, and I have a comment block auto-complete plugin for VSCode.

File: src/auth/pages/Login/Login.tsx (hunk 4 above)
Comment (eliasmalik) [reply src/auth/pages/Login/Login.tsx:38]: We already have `redirectOnError`, but we need to handle these three cases separately anyway. I can use it in the final case.

File: src/auth/components/PrivateRoute.tsx (hunk 2 above)
Comment (eliasmalik) [reply src/auth/components/PrivateRoute.tsx:25]: Yeah lets talk about this, but IMO out of scope of this PR. 

For reference, this is the equivalent implementation with hooks:
```tsx
const PrivateRoute: React.SFC<PrivateRouteProps> = (props) => {
  const { component: Comp, ...rest } = props;
  const [authStatus, setAuthStatus] = useState(AuthStatusEnum.PENDING);

  useEffect(() => {
    Roles.get()
      .then(() => setAuthStatus(AuthStatusEnum.SUCCESS))
      .catch(() => setAuthStatus(AuthStatusEnum.FAILURE));
  });

  switch (authStatus) {
    case AuthStatusEnum.PENDING:
      return (
        <FullScreenBeatLoader color={colors.highlight_primary} />
      );

    case AuthStatusEnum.SUCCESS:
      return (
        <Route {...rest} component={Comp} />
      );

    case AuthStatusEnum.FAILURE:
    default:
      return (
        <Redirect to={{ pathname: '/login' }} />
      );
  }
};

export default PrivateRoute;
```

**Review by astroash — APPROVED** [review]

**Pull request merged**